### PR TITLE
Fix bug due to rounding errors in periodicity computation

### PR DESCRIPTION
**Pull request by efaulhaber**
On main, we first computed the periodic continuous coordinates and then converted to discrete cell coordinates. The first computation introduces rounding errors, sometimes resulting in coordinates outside the periodic box, which will then be converted to discrete cell coordinates outside the box as well. This leads to out-of-bounds errors.
This PR first computes the discrete cell coordinates and then applies a discrete modulo, so there are no rounding errors that can break anything.

### CODE REVIEW

**Review by efaulhaber — COMMENTED** [review]
## Codex Review

The PR appears intended to centralize periodic cell-index wrapping and make FullGridCellList use the same periodic path as other cell lists.

_This is an AI-generated code review. Please verify the findings and summary before acting on them._
<sub>Review generated by [codex-pr-review](https://github.com/efaulhaber/codex-pr-review)</sub>

File: test/nhs_grid.jl [hunk 1]
```diff
@@ -239,7 +238,7 @@
             [-0.08 0.0 0.18 0.1 -0.08
              -0.12 -0.05 -0.09 0.15 0.39],
             [-0.08 0.0 0.18 0.1 -0.08
-             -0.12 -0.05 -0.09 0.15 0.42],
+             -0.12 -0.05 -0.09 0.15 0.42] .- [0.0017, 0.005],
```
Comment (efaulhaber) [inline test/nhs_grid.jl:241]: **Medium:** The updated periodic fixture shifts both coordinates and the periodic box, so it no longer covers the non-origin/unaligned periodic-box case most likely to catch regressions in periodic index handling. Keep or add an unshifted non-origin fixture, including FullGridCellList, with expected neighbor assertions.

File: src/nhs_grid.jl [hunk 2]
```diff
@@ -586,37 +586,44 @@ end
...
 end
 
-@inline periodic_cell_index(cell_index, ::Nothing, n_cells, cell_list) = cell_index
-
-@inline function periodic_cell_index(cell_index, ::PeriodicBox, n_cells, cell_list)
-    # 1-based modulo
-    return mod1.(cell_index, n_cells)
+@inline periodic_cell_index(cell_index, ::Nothing, n_cells) = cell_index
+
+@inline function periodic_cell_index(cell_index, ::PeriodicBox, n_cells)
```
Comment (efaulhaber) [inline src/nhs_grid.jl:596]: **Low:** The generic periodic_cell_index now hard-codes the FullGridCellList padding offset even though it no longer dispatches on the cell-list type. Move this offset back behind cell-list-specific dispatch or make it part of the cell-list abstraction so future layouts do not need to preserve this hidden convention.

File: src/cell_lists/full_grid.jl [hunk 3]
```diff
@@ -81,33 +81,11 @@ function FullGridCellList(; min_corner, max_corner,
...
-end
-
-@inline function periodic_cell_index(cell_index, ::PeriodicBox, n_cells,
-                                     cell_list::FullGridCellList)
-    # 2-based modulo to match the indexing of the periodic box explained above.
-    return mod.(cell_index .- 2, n_cells) .+ 2
+@inline function nonperiodic_cell_coords(coords, cell_list::FullGridCellList, cell_size)
+    # Subtract `min_corner` to offset coordinates so that the (padded) min corner
+    # of the grid corresponds to the (1, 1, 1) cell.
+    # The unpadded min corner (that is passed by the user) corresponds to the (2, 2, 2) cell.
```
Comment (efaulhaber) [inline src/cell_lists/full_grid.jl]: **Low:** The comment says the user-provided min corner maps to cell (2, 2, 2), but that is not necessarily true when periodic search rounds cell_size above search_radius. Qualify the comment to the unrounded/default case or describe the padding intent without promising an exact cell index.

File: test/nhs_grid.jl (hunk 1 above)
Comment (efaulhaber) [inline test/nhs_grid.jl:241]: **Low:** The alignment offset vector is repeated in both the coordinates and periodic box setup, making the fixture easy to desynchronize. Introduce a named local such as alignment_offset and reuse it in both places.

File: test/neighborhood_search.jl [hunk 4]
```diff
@@ -1,6 +1,39 @@
...
+                      SVector(nextfloat(T(1)), T(0.5)),
+                      SVector(T(0.5), prevfloat(T(0))),
+                      SVector(T(0.5), T(0)),
+                      SVector(T(0.5), nextfloat(T(0))),
+                      SVector(T(0.5), prevfloat(T(1))),
+                      SVector(T(0.5), T(1)),
+                      SVector(T(0.5), nextfloat(T(1))))
+
+            for x in coords
+                xp = PointNeighbors.periodic_coords(x, box)
```
Comment (efaulhaber) [inline test/neighborhood_search.jl:24]: **Low:** These tests only exercise `periodic_coords` directly, but the changed Grid/FullGrid `cell_coords` path can regress when actual neighbor searches use boundary inputs such as `prevfloat(min_corner)` or `nextfloat(max_corner)`. Add a small periodic Grid/FullGrid neighbor-search test at those boundary inputs and assert the expected neighbors.

File: src/nhs_grid.jl (hunk 2 above)
Comment (efaulhaber) [inline src/nhs_grid.jl:596]: **Low:** This generic periodic-indexing path now hard-codes the FullGridCellList-style `2` offset, which weakens the existing cell-list abstraction and makes future cell lists with different padding/indexing semantics modify shared grid code. Keep `cell_list` in the dispatch or move this offset logic behind a cell-list-specific method.

**Review by efaulhaber — COMMENTED** [review]
## Codex Review

No correctness or regression issues were identified in the changed production behavior for this pass.

_This is an AI-generated code review. Please verify the findings and summary before acting on them._
<sub>Review generated by [codex-pr-review](https://github.com/efaulhaber/codex-pr-review)</sub>

File: src/cell_lists/full_grid.jl (hunk 3 above)
Comment (svchb) [inline src/cell_lists/full_grid.jl]:
```suggestion
# Subtract `min_corner` to account for the padded finite grid.
# In the common unrounded case, the user-provided min corner maps to cell 2,
# leaving one padding layer for neighbor-cell bounds checks.
```

File: test/neighborhood_search.jl [hunk 5]
```diff
@@ -1,6 +1,39 @@
...
+
+                # Test that the periodic coordinates are within the periodic box.
+                @test all(box.min_corner .<= xp)
+                @test all(xp .<= box.max_corner)
+
+                # Test that the periodic coordinates are equivalent to
+                # the original coordinates up to periodicity.
+                @test xp[1] in (x[1], x[1] - box.size[1], x[1] + box.size[1])
+                @test xp[2] in (x[2], x[2] - box.size[2], x[2] + box.size[2])
+            end
```
Comment (svchb) [inline test/neighborhood_search.jl:34]:
```suggestion
            end
            
              # Also exercise the actual grid/cell-list path for boundary-rounding inputs.
            search_radius = T(0.25)
            boundary_coords = hcat(SVector(prevfloat(T(0)), T(0.5)),
                                   SVector(prevfloat(T(1)), T(0.5)))

            cell_list = FullGridCellList(; min_corner = box.min_corner,
                                         max_corner = box.max_corner,
                                         search_radius)

            nhs = GridNeighborhoodSearch{2}(; search_radius,
                                            n_points = size(boundary_coords, 2),
                                            periodic_box = box,
                                            cell_list)

            initialize!(nhs, boundary_coords, boundary_coords)

            neighbors = [Int[] for _ in axes(boundary_coords, 2)]
            foreach_point_neighbor(boundary_coords, boundary_coords, nhs) do point, neighbor,
                                                                              pos_diff, distance
                push!(neighbors[point], neighbor)
            end
         
            
            
```

**Review by svchb — CHANGES_REQUESTED** [review]

**Review by svchb — APPROVED** [review]
